nvflare/fuel/f3/drivers/base_driver.py

```diff
@@ -0,0 +1,66 @@
+# Copyright (c) 2021-2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import threading
+from abc import ABC
+from typing import Dict, Optional
+
+from nvflare.fuel.f3.connection import Connection, ConnState
+from nvflare.fuel.f3.drivers.driver import ConnectorInfo, Driver
+
+log = logging.getLogger(__name__)
+
+
+class BaseDriver(Driver, ABC):
+    """Common base class for all drivers
+    It contains all the common connection management code
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.connections: Dict[str, Connection] = {}
+        self.connector: Optional[ConnectorInfo] = None
+        self.conn_lock = threading.Lock()
+
+    def add_connection(self, conn: Connection):
+        with self.conn_lock:
+            self.connections[conn.name] = conn
+
+        conn.state = ConnState.CONNECTED
+        self._notify_monitor(conn)
+
+        log.debug(f"Connection created: {self.get_name()}:{conn}")
+
+    def close_connection(self, conn: Connection):
+        log.debug(f"Connection removed: {self.get_name()}:{conn}")
+
+        conn.state = ConnState.CLOSED
+        self._notify_monitor(conn)
+
+        with self.conn_lock:
+            if not self.connections.pop(conn.name, None):
+                log.debug(f"{conn.name} is already removed from driver")
```

Comment:
Can the connection name be `""`?

In that case should we use `if self.connections.pop(conn.name, None) is not None:`?

And should we use `log.debug(f"{conn.name} is successfully removed from driver")` ?

Reply:
Connection name can't be None.

"Already removed" is a corner case so I want that printed instead of the normal case.

Reply:
In that case,
should the logic be

```
           if self.connections.pop(conn.name, None) is None:
                log.debug(f"{conn.name} is already removed from driver")
```

Reply:
"not x" and "x is None" should be identical in this use case. The former is more Pythonic.

Reply:
IMO here (since we use `pop(something, None)`) use `is None` or `is not None` is more Pythonic, but I am OK with `not x`